File: problems/hackerrank/arrays/bomberMan/bomberMan.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// bomberMan simulates the Bomberman game over n seconds.
func bomberMan(n int32, grid []string) []string {
	r := len(grid)
	c := len(grid[0])

	if n == 1 {
		return grid
	}

	if n%2 == 0 {
		var fullGrid []string
		row := strings.Repeat("O", c)
		for i := 0; i < r; i++ {
			fullGrid = append(fullGrid, row)
		}
		return fullGrid
	}

	detonate := func(g []string) []string {
		newGrid := make([][]rune, r)
		for i := range newGrid {
			newGrid[i] = make([]rune, c)
			for j := range newGrid[i] {
				newGrid[i][j] = 'O'
			}
		}

		for i := 0; i < r; i++ {
			for j := 0; j < c; j++ {
				if g[i][j] == 'O' {
					newGrid[i][j] = '.'
					if i > 0 {
						newGrid[i-1][j] = '.'
					}
					if i < r-1 {
						newGrid[i+1][j] = '.'
					}
					if j > 0 {
						newGrid[i][j-1] = '.'
					}
					if j < c-1 {
						newGrid[i][j+1] = '.'
					}
				}
			}
		}

		var res []string
		for i := 0; i < r; i++ {
			res = append(res, string(newGrid[i]))
		}
		return res
	}

	grid3 := detonate(grid)
	if n%4 == 3 {
		return grid3
	}
	grid5 := detonate(grid3)
	return grid5
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	firstMultipleInput := strings.Split(strings.TrimSpace(readLine(reader)), " ")

	rTemp, err := strconv.ParseInt(firstMultipleInput[0], 10, 64)
	checkError(err)
	r := int32(rTemp)

	cTemp, err := strconv.ParseInt(firstMultipleInput[1], 10, 64)
	checkError(err)
	_ = int32(cTemp)

	nTemp, err := strconv.ParseInt(firstMultipleInput[2], 10, 64)
	checkError(err)
	n := int32(nTemp)

	var grid []string

	for i := 0; i < int(r); i++ {
		gridItem := readLine(reader)
		grid = append(grid, gridItem)
	}

	result := bomberMan(n, grid)

	for i, rowItem := range result {
		fmt.Printf("%s", rowItem)
		if i != len(result)-1 {
			fmt.Printf("\n")
		}
	}
	fmt.Printf("\n")
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == nil {
		return string(str)
	}
	return ""
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}

File: problems/hackerrank/arrays/bomberMan/bomberMan.py
#!/bin/python3

import math
import os
import random
import re
import sys

def bomberMan(n, grid):
    """
    Simulate the Bomberman game over n seconds.
    
    Bomberman lives in a rectangular grid. Each cell is empty or contains a bomb.
    Initially, he plants bombs in some cells.
    After 1 second, he does nothing.
    After one more second, he plants bombs in all empty cells.
    After one more second, bombs planted 3 seconds ago detonate.
    This pattern repeats.
    
    Args:
        n (int): The number of seconds to simulate.
        grid (list of str): A list of strings representing the grid where 'O' is a bomb and '.' is an empty cell.
        
    Returns:
        list of str: The state of the grid after `n` seconds.
    """
    r = len(grid)
    c = len(grid[0])

    if n == 1:
        return grid

    elif n % 2 == 0:
        return ['O' * c for _ in range(r)]

    else:
        bomb_neighbors = []
        for row in range(r):
            for column in range(c):
                if grid[row][column] == 'O':
                    bomb_neighbors.append((row, column))
                    up = (row - 1, column)
                    down = (row + 1, column)
                    left = (row, column - 1)
                    right = (row, column + 1)
                    for neighbor in [up, down, left, right]:
                        if 0 <= neighbor[0] < r and 0 <= neighbor[1] < c:
                            bomb_neighbors.append(neighbor)

        grid = [['O' for _ in range(c)] for _ in range(r)]
        for row, column in bomb_neighbors:
            grid[row][column] = '.'

        if n % 4 == 3:
            return [''.join(row) for row in grid]

        bomb_neighbors = []
        for row in range(r):
            for column in range(c):
                if grid[row][column] == 'O':
                    bomb_neighbors.append((row, column))
                    up = (row - 1, column)
                    down = (row + 1, column)
                    left = (row, column - 1)
                    right = (row, column + 1)
                    for neighbor in [up, down, left, right]:
                        if 0 <= neighbor[0] < r and 0 <= neighbor[1] < c:
                            bomb_neighbors.append(neighbor)

        grid = [['O' for _ in range(c)] for _ in range(r)]
        for row, column in bomb_neighbors:
            grid[row][column] = '.'

        return [''.join(row) for row in grid]

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    first_multiple_input = input().rstrip().split()

    r = int(first_multiple_input[0])

    c = int(first_multiple_input[1])

    n = int(first_multiple_input[2])

    grid = []

    for _ in range(r):
        grid_item = input()
        grid.append(grid_item)

    result = bomberMan(n, grid)

    fptr.write('\n'.join(result))
    fptr.write('\n')

    fptr.close()